docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,345 @@
+
+The `length` parameter:
+
+- Specifies how many bits to compare.
+- Adds more constraints for larger numbers.
+
+For details about the implementation, see [XOR](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#xor-1) in the Mina book.
+
+Example:
+
+```ts
+let a = Field(0b0101);
+let b = Field(0b0011);
+
+let c = Gadgets.xor(a, b, 4); // xor-ing 4 bits
+c.assertEquals(0b0110);
+```
+
+## leftShift32()
+
+```ts
+leftShift32(field: Field, bits: number) => Field
+```
+
+The `leftShift32()` gadget is a provable equivalent to the [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operator in JavaScript. It moves the bits of a binary number to the left by the specified number of `bits`. Any bits that fall off the left side are discarded. `0`s are padded in from the right. It returns a new `Field` element that is range checked to 32 bits.
+
+The input `Field` should not 32 bits in size. You can use [rangeCheck32](#rangecheck32) to ensure this.
+
+Example:
+
+```ts
+const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
+const y = Gadgets.leftShift32(x, 2); // left shift by 2 bits
+y.assertEquals(0b110000); // 48 in binary
+```
+
+## leftShift64()
+
+```ts
+leftShift64(field: Field, bits: number) => Field
+```
+
+The `leftShift64()` gadget is a provable equivalent to the [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operator in JavaScript. It moves the bits of a binary number to the left by the specified number of `bits`. Any bits that fall off the left side are discarded. `0`s are padded in from the right. It returns a new `Field` element that is range checked to 64 bits.
+
+The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.
+
+Example:
+
+```ts
+const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
+const y = Gadgets.leftShift64(x, 2); // left shift by 2 bits
+y.assertEquals(0b110000); // 48 in binary
+
+const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
+Gadgets.leftShift64(xLarge, 32); // throws an error since input exceeds 64 bits
+```
+
+## rightShift64()
+
+```ts
+rightShift64(field: Field, bits: number) => Field
+```
+
+The `rightShift64()` gadget is a provable equivalent to the [Right shift (>>)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift) operator in JavaScript. It moves the bits of a binary number to the right by the specified number of `bits`. Any bits that fall off the right side are discarded. `0`s are padded in from the left. It returns a new `Field` element.
+
+The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.
```

Comment:
```suggestion
The input `Field` must not exceed 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.
```

same comment as above
"must not" or "cannot" (but not should)

let's add "exceeds"